Separate async sync api

chrisblossom opened the pull request:
This PR separates the `async` and `sync` apis from each other as described in #209.

```js
import { cosmiconfig, cosmiconfigSync } from 'cosmiconfig';

const explorer = cosmiconfig({
	loaders: {
		'.js': async () => {},
		'.ts': () => {}, // can be sync
	},
	transform: async (config) => config,
	// transform: (config) => config, // can also be sync
});

await explorer.search();
await explorer.load();

const explorerSync = cosmiconfigSync({
	loaders: {
		'.js': () => {},
		'.ts': async () => {}, // can not be async, invalid
	},
	transform: (config) => config,
	// transform: async (config) => config, // can not be async, invalid
});

explorerSync.search();
explorerSync.load();
```

I think the readme could be updated further to be more condensed and clear by adding a note about the `cosmiconfigSync` separation and removing all references to what was previously `loadSync` and `searchSync`.

davidtheclark commented:
Nice work, @chrisblossom. I've been on vacation. I'll try to review this within the next few days so we can keep moving towards v6.

davidtheclark commented on src/createExplorerBase.ts:
```diff
@@ -0,0 +1,146 @@
+} from './types';
+import { Loader } from './index';
+
+class ExplorerBase<T extends ExplorerOptions | ExplorerOptionsSync> {
```
Curious: why is this `T extends A | B` rather than `A | B`?

chrisblossom replied:
Generics are used here because `ExplorerOptions` are not compatible with `ExplorerOptionsSync` so you have to explicitly tell typescript what object it is expecting so it can return the correct type. At least that is how I understand it.

If you try removing the generic and replacing it with `ExplorerOptions | ExplorerOptionsSync` You'll see several errors in `createExplorerSync.ts`.

davidtheclark commented on src/createExplorerBase.ts:
```diff
@@ -0,0 +1,146 @@
+  return ext ? `extension "${ext}"` : 'files without extensions';
+}
+
+export { ExplorerBase, getExtensionDescription };
```
Should we change this filename to `ExplorerBase` since it's no longer exporting a factory function?

chrisblossom replied:
Sounds good to me. I think we should also rename the class files and move the factory creation into `index.ts`. How does that sound?

davidtheclark replied:
Sure.

davidtheclark commented on test/index.test.ts:
```diff
@@ -1,14 +1,17 @@
+let createExplorerSyncMock = jest.spyOn(
+  createExplorerSync,
+  'createExplorerSync',
+);
```
Does `jest.mock` not work to accomplish the equivalent?

chrisblossom replied:
`jest.mock` removes the implementation of the file, while `jest.spyOn` does not (by default) which then tests the application instead of the mock we created.

davidtheclark commented on test/index.test.ts:
```diff
@@ -1,14 +1,17 @@
+let createExplorerSyncMock = jest.spyOn(
+  createExplorerSync,
+  'createExplorerSync',
+);
```
Do we need to `spyOn` here if we're doing it in the `beforeEach` hook?

chrisblossom replied:
For testing, no. I added it so typescript can infer the correct type. Maybe there is a better way to do it, but I am unsure how?

davidtheclark commented on README.md:
```diff
@@ -364,9 +366,7 @@ Do whatever you need to, and return either a configuration object or `null` (or,
-Then it can be used regardless of whether you end up calling [`search()`] or [`searchSync()`], [`load()`] or [`loadSync()`].
-If you want or need to provide an async-only loader, you can do so by making the value of `loaders` an object with an `async` property whose value is the async loader.
-You can also add a `sync` property to designate a sync loader, if you want to use both async and sync search and load functions.
+Then it can be used regardless of whether you end up calling [`search()`], [`load()`].
```
This is no longer accurate, right?

chrisblossom commented:
Sorry for the delay, been very busy lately but I should have had this done already. I am unsure if you're wanting me to push directly to this branch, let me know if you don't want me to.

I think I addressed all issues we discussed in #211.

I do not particularly like how I fixed the spy instances in `index.test.ts`. I also ran into a strange jest issue where `toMatchObject` was not working with `transform: expect.any(function)`. If someone has a better solution I'm all ears. I also think they could be completely removed as they don't really test anything important.
